lib/client.js

```diff
@@ -2222,6 +2222,123 @@ MatrixClient.prototype.deletePushRule = function(scope, kind, ruleId, callback)
+                                                     ruleId, enabled, callback) {
+    var path = utils.encodeUri("/pushrules/" + scope + "/$kind/$ruleId/enabled", {
+        $kind: kind,
+        $ruleId: ruleId
+    });
+    return this._http.authedRequest(
+        callback, "PUT", path, undefined, enabled ? 'true' : 'false'
+    );
+};
+
+/**
+ * Get the room-kind push rule associated with a room.
+ * @param {string} scope "global" or device-specific.
+ * @param {string} roomId the id of the room.
+ * @return {object} the rule or undefined.
+ */
+MatrixClient.prototype.getRoomPushRule = function(scope, roomId) {
+    // There can be only room-kind push rule per room
+    // and its id is the room id.
+    if (this.pushRules) {
+        for (var i = 0; i < this.pushRules[scope].room.length; i++) {
+            var rule = this.pushRules[scope].room[i];
+            if (rule.rule_id === roomId) {
+                return rule;
+            }
+        }
+    }
+};
+
+/**
+ * Set a room-kind muting push rule in a room.
+ * The operation also updates MatrixClient.pushRules at the end.
+ * @param {string} scope "global" or device-specific.
+ * @param {string} roomId the id of the room.
+ * @param {string} mute the mute state.
+ * @return {module:client.Promise} Resolves: result object
+ * @return {module:http-api.MatrixError} Rejects: with an error response.
+ */
+MatrixClient.prototype.setRoomMutePushRule = function(scope, roomId, mute) {
+    var self = this;
+    var deferred, hasDontNotifyRule;
+
+    // Get the existing room-kind push rule if any
+    var roomPushRule = this.getRoomPushRule(scope, roomId);
+    if (roomPushRule) {
+        if (0 <= roomPushRule.actions.indexOf("dont_notify")) {
+            hasDontNotifyRule = true;
+        }
+    }
+
+    if (!mute) {
+        // Remove the rule only it is a muting rule
+        if (hasDontNotifyRule) {
+            deferred = this.deletePushRule(scope, "room", roomPushRule.rule_id);
+        }
+    }
+    else {
+        if (!roomPushRule) {
+            deferred = this.addPushRule(scope, "room", roomId, {
+                actions: ["dont_notify"]
+            });
+        }
+        else if (!hasDontNotifyRule) {
+            // Remove the existing one before setting the mute push rule
+            var deferred2 = q.defer();
+            this.deletePushRule(scope, "room", roomPushRule.rule_id)
+            .done(function() {
+                self.addPushRule(scope, "room", roomId, {
+                    actions: ["dont_notify"]
+                }).done(function() {
+                    deferred2.resolve();
+                }, function(err) {
+                    deferred2.reject(err);
+                });
+            }, function(err) {
+                deferred2.reject(err);
+            });
+
+            deferred = deferred2.promise;
+        }
+    }
+
+    if (deferred) {
+        // Update this.pushRules when the operation completes
+        var deferred3 = q.defer();
```

Comment:
I'm a bit lost with 3 different deferreds here. Maybe this could be rule_refresh_deferred or something?